extensions/_modules/heroku.py

```diff
@@ -51,164 +46,225 @@ def __virtual__():
-  :return:            The json response from the API call or False.
-  '''
-  headers = {}
-  query_params = {}
-
-  if not api_url:
-      try:
-          options = __salt__['config.option']('heroku')
-          api_url = options.get('api_url')
-      except (NameError, KeyError, AttributeError):
-          pass  # not mandatory, thus won't fail if not found
-
-  if not api_key or not endpoint:
-      try:
-          options = __salt__['config.option']('heroku')
-          if not api_key:
-            api_key = options.get('api_key')
-          if not endpoint:
-            endpoint = options.get('endpoint')
-      except (NameError, KeyError, AttributeError):
-          log.error("No Heroku api key or endpoint found.")
-          return False
-
-  use_api_url = 'https://api.heroku.com'  # default API URL
-  if api_url:
-      use_api_url = api_url
-  url = _urljoin(use_api_url, endpoint + '/')
-  if arguments:
-    url = _urljoin(url, arguments + '/')
-
-  if endpoint:
-    headers['Authorization'] = 'Bearer {0}'.format(api_key)
-    headers['Accept'] = 'application/vnd.heroku+json; version=3'
-    if data:
-      data = json.dumps(data)
-
-    if method == 'POST' or method == 'PATCH':
-      headers['Content-Type'] = 'application/json'
-  else:
-    log.error('Heroku endpoint not specified')
-    return False
-
-  result = salt.utils.http.query(
-      url,
-      method,
-      params=query_params,
-      data=data,
-      decode=True,
-      status=True,
-      header_dict=headers,
-      opts=__opts__,
-  )
-
-  if result.get('status', None) == salt.ext.six.moves.http_client.OK:
-      return result.get('dict', {})
-  elif result.get('status', None) == salt.ext.six.moves.http_client.NO_CONTENT:
-      return False
-  else:
-      log.debug(url)
-      log.debug(query_params)
-      log.debug(data)
-      log.debug(result)
-      if result.get('error'):
-          log.error(result)
-      return False
-
-def list_apps(api_url=None,
-               api_key=None,
-               endpoint=None):
+def _query(api_key=None, endpoint=None, arguments=None, method='GET', data=None):
+    '''
+    Heroku object method function to construct and execute on the API URL.
+
+    :param api_key:     The Heroku api key.
+    :param endpoint:    The Heroku api function to perform.
+    :param arguments:   The Heroku arguments to pass.
+    :param method:      The HTTP method, e.g. GET, POST, or PATCH.
+    :param data:        The data to be sent for POST/PATCH method.
+    :return:            The json response from the API call or False.
+    '''
+    headers = {}
+    query_params = {}
+
+    if not api_key or not endpoint:
+        try:
+            options = __salt__['config.option']('heroku')
+            if not api_key:
+                api_key = options.get('api_key')
+            if not endpoint:
+                endpoint = options.get('endpoint')
+        except (NameError, KeyError, AttributeError):
```

Comment:
The `get` function will never throw an exception and will instead return `None` as the default value. If the intent of the `except` clause is to handle the situation where there is no `heroku` configuration parameter then it may be useful to update the log message to reflect that, as well as passing that through as part of the return value.